### ISSUE TITLE
File Descriptor Leak

### ISSUE THREAD

**Issue by inklit** [SquidDev-CC/CC-Tweaks#118]
This error doesn't particularly point to CC as the issue however I can't help but assume that it is at fault here since it was the only thing suffering from the consequences of this problem.

```
java.io.IOException: Cannot run program "sh": error=24, Too many open files
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
        at jline.internal.TerminalLineSettings.exec(TerminalLineSettings.java:308)
        at jline.internal.TerminalLineSettings.stty(TerminalLineSettings.java:282)
        at jline.internal.TerminalLineSettings.undef(TerminalLineSettings.java:158)
        at jline.UnixTerminal.disableLitteralNextCharacter(UnixTerminal.java:185)
        at jline.console.ConsoleReader.readLine(ConsoleReader.java:2448)
        at jline.console.ConsoleReader.readLine(ConsoleReader.java:2372)
        at jline.console.ConsoleReader.readLine(ConsoleReader.java:2360)
        at net.minecraftforge.server.console.TerminalHandler.handleCommands(TerminalHandler.java:36)
        at net.minecraft.server.dedicated.DedicatedServer$2.run(DedicatedServer.java:86)
Caused by: java.io.IOException: error=24, Too many open files
        at java.lang.UNIXProcess.forkAndExec(Native Method)
        at java.lang.UNIXProcess.<init>(UNIXProcess.java:248)
        at java.lang.ProcessImpl.start(ProcessImpl.java:134)
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
        ... 9 more
```

Essentially, there is a file descriptor leak somewhere. It keeps too many file handles open and after a while the server is NOT happy. This prevents ComputerCraft from opening/saving files, but it also **prevents the world from saving, and loading old chunks.**

Restarting the server fixes the issue - but while this issue is in effect ComputerCraft cannot access any files and really flips out.

A way to reproduce the issue:

```
while true do
    fs.open("test", w)
end
```

Simply put, reduction of bad programming would fix this issue however I run a public server and can't really enforce or moderate people's ability to program well.

My suggested solution is to add a file descriptor limit for every individual computer and close file descriptors on reboot/shutdown of the computer automatically.

**Comment by SquidDev** [comment]
I'm going to add a limit in CCTweaks-Lua but may I recommend you also suggest this on the ComputerCraft forums/bug tracker - I'd rather this was in ComputerCraft core.

**Comment by SquidDev** [comment]
Fixed in [CCTweaks-Lua](https://github.com/SquidDev-CC/CCTweaks-Lua/commit/722a8b67eaf051c562e3706bb04e37f2a6b82fa4). Do you think you could also post this on the CC issue tracker? I don't want to have to add this hack.

**Comment by inklit** [comment]
Thanks for this!  Just wondering, if I compile CCTweaks from source, will it also get the additions from CCTweaks-Lua?

**Comment by SquidDev** [comment]
If you bump the CCTweaks-Lua version to 1.0.11 then it should work correctly. I'm hoping to fix #117 then I'll push a release.

**Comment by SquidDev** [comment]
Actually, scratch that: it appears everything is broken.

**Comment by SquidDev** [comment]
Apparently 1.0.11 is fine to use, just Travis/Gradle is broken.
